### ISSUE TITLE
"unused macro definition" false positive with a private helper macro

### ISSUE THREAD

**Issue by SillyFreak** [rust-lang/rust#61265]
consider the following:

```rust
macro_rules! internal {
    () => { };
}

#[macro_export]
macro_rules! public {
    () => { internal!() };
}

pub fn user() {
    // public!();
}
```

This code gives me `unused macro definition` for the `internal!` macro. Of course, if I were to remove it, the exported `public!` macro would not work. `internal!` is unused in a very narrow sense, but it's a false positive for sure.

**Comment by SillyFreak** [comment]
ok, trying to use the macro from another module I just noticed that I seem to have bot macros exported to have this actually work correctly.

Is that correct, or am I missing even more?

A workaround for both seems to be to declare the helper macro inside the exported macro. Is that how it's intended?

**Comment by petrochenkov** [comment]
>ok, trying to use the macro from another module I just noticed that I seem to have bot macros exported to have this actually work correctly.

Yes, `internal` will be searched at the call site of `public`, so it has to be public enough to be available from there.

Something like this is usually used for public macros with internal helpers
```rust
#[doc(hidden)]
#[macro_export]
macro_rules! internal {
    () => { };
}

#[macro_export]
macro_rules! public {
    () => { $crate::internal!() };
}
```

**Comment by petrochenkov** [comment]
So, in the original examle the "unused" warning is working as intended, closing the issue.

**Comment by SillyFreak** [comment]
thanks for the clarifications!
